File: Leetcode - Java/1602_tree_findNearestRightNode.java
// https://leetcode.com/problems/find-nearest-right-node-in-binary-tree/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode findNearestRightNode(TreeNode root, TreeNode u) {
        if (root == null) return null;
        
        Queue<TreeNode> q = new LinkedList();
        TreeNode curr = null;
        q.offer(root);
        q.offer(null);
        
        while (!q.isEmpty()) {
            curr = q.poll();
            
            if (curr != null) {
                if (curr == u) return q.poll();
                if (curr.left != null) q.offer(curr.left);
                if (curr.right != null) q.offer(curr.right);
            } else {
                if (!q.isEmpty()) q.offer(null);
            }
        }
        return null;
    }
}

File: Leetcode - Python/1602_tree_findNearestRightNode.py
# https://leetcode.com/problems/find-nearest-right-node-in-binary-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque


class Solution:
    def findNearestRightNode(self, root: TreeNode, u: TreeNode) -> TreeNode:
        if root is None:
            return None

        q = deque([root, None, ])
        while q:
            curr = q.popleft()

            if curr == u:
                return q.popleft()

            if curr:
                if curr.left:
                    q.append(curr.left)
                if curr.right:
                    q.append(curr.right)

            else:
                if q:
                    q.append(None)
